src/app/utils/rpc-utils.ts

```diff
@@ -0,0 +1,67 @@
+import { JsonRpcProvider, formatUnits } from 'ethers'
+import { Network } from '../../types/network'
+import { paraTimesConfig } from '../../config'
+import { Layer } from '../../oasis-nexus/api'
+
+const RPC_ENDPOINTS: Partial<Record<Network, Partial<Record<Layer, string>>>> = {
+  [Network.mainnet]: {
+    [Layer.emerald]: 'https://emerald.oasis.dev',
+    [Layer.sapphire]: 'https://sapphire.oasis.io',
+  },
+  [Network.testnet]: {
+    [Layer.emerald]: 'https://testnet.emerald.oasis.dev',
+    [Layer.sapphire]: 'https://testnet.sapphire.oasis.dev',
+  },
+}
+
+const NETWORK_DECIMALS: Partial<Record<Layer, number>> = {
+  [Layer.emerald]: paraTimesConfig[Layer.emerald].decimals,
+  [Layer.sapphire]: paraTimesConfig[Layer.sapphire].decimals,
+  [Layer.cipher]: paraTimesConfig[Layer.cipher].decimals,
+}
+
+interface RpcProviderOptions {
+  context: {
+    network: Network
+    layer: Layer
+  }
+}
+
+export abstract class RpcUtils {
+  private static _provider: JsonRpcProvider | null = null
+  private static _providerLayer: Layer | null = null
+  private static _providerNetwork: Network | null = null
+
+  private static _getProvider(opts: RpcProviderOptions) {
+    const {
+      context: { network, layer },
+    } = opts
+    const rpcEndpoint = RPC_ENDPOINTS[network]?.[layer]
+
+    if (!rpcEndpoint) {
+      throw new Error('Invalid scope, unable to create Provider!')
+    }
+
+    if (RpcUtils._providerNetwork === network && RpcUtils._providerLayer === layer && RpcUtils._provider) {
+      return RpcUtils._provider
+    }
+
+    RpcUtils._provider = new JsonRpcProvider(rpcEndpoint)
+    RpcUtils._providerNetwork = network
+    RpcUtils._providerLayer = layer
+
+    return RpcUtils._provider
+  }
+
+  static async getAccountBalance(address: string, opts: RpcProviderOptions): Promise<string> {
+    const provider = RpcUtils._getProvider(opts)
+
+    const balance = await provider.getBalance(address)
+
+    const {
+      context: { layer },
+    } = opts
+    const decimals = NETWORK_DECIMALS[layer] ?? 18
```

Comment:
I would remove the `?? 18` fallback and just add `consensus: 9` to layer decimals